File: app/src/test/java/com/coinninja/coinkeeper/service/runner/BroadcastBtcInviteRunnerTest.java
package com.coinninja.coinkeeper.service.runner;

import android.content.Context;
import android.content.res.Resources;

import com.coinninja.bindings.DerivationPath;
import com.coinninja.bindings.TransactionBroadcastResult;
import com.coinninja.bindings.TransactionData;
import com.coinninja.bindings.UnspentTransactionOutput;
import com.coinninja.coinkeeper.bitcoin.BroadcastTransactionHelper;
import com.coinninja.coinkeeper.cn.transaction.TransactionNotificationManager;
import com.coinninja.coinkeeper.cn.wallet.SyncWalletManager;
import com.coinninja.coinkeeper.cn.wallet.tx.TransactionFundingManager;
import com.coinninja.coinkeeper.model.PaymentHolder;
import com.coinninja.coinkeeper.model.PhoneNumber;
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary;
import com.coinninja.coinkeeper.model.db.TransactionsInvitesSummary;
import com.coinninja.coinkeeper.model.helpers.BroadcastBtcInviteHelper;
import com.coinninja.coinkeeper.model.helpers.ExternalNotificationHelper;
import com.coinninja.coinkeeper.model.helpers.InviteTransactionSummaryHelper;
import com.coinninja.coinkeeper.model.helpers.TransactionHelper;
import com.coinninja.coinkeeper.model.helpers.WalletHelper;
import com.coinninja.coinkeeper.util.analytics.Analytics;
import com.coinninja.coinkeeper.util.currency.BTCCurrency;
import com.coinninja.coinkeeper.util.currency.USDCurrency;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class BroadcastBtcInviteRunnerTest {
    @Mock
    private Context context;
    @Mock
    private TransactionFundingManager transactionFundingManger;
    @Mock
    private Resources resources;
    @Mock
    private InviteTransactionSummary invite;
    @Mock
    private Analytics analytics;
    @Mock
    private PaymentHolder paymentHolder;
    @Mock
    private BroadcastBtcInviteHelper broadcastBtcInviteHelper;
    @Mock
    private TransactionData transactionData;
    @Mock
    private TransactionHelper transactionHelper;
    @Mock
    private ExternalNotificationHelper externalNotificationHelper;
    @Mock
    private WalletHelper walletHelper;
    @Mock
    private BroadcastTransactionHelper broadcastTransactionHelper;
    @Mock
    private InviteTransactionSummaryHelper inviteTransactionSummaryHelper;
    @Mock
    private TransactionNotificationManager transactionNotificationManager;
    @Mock
    private TransactionBroadcastResult transactionBroadcastResult;
    @Mock
    private TransactionsInvitesSummary transactionsInvitesSummary;
    @Mock
    private SyncWalletManager syncWalletManager;
    private String serverId = "--server-id--";
    @Mock
    private PhoneNumber phoneNumber;

    @InjectMocks
    private BroadcastBtcInviteRunner runner;
    String address = "--address--";

    @After
    public void tearDown() {
        runner = null;
        address = null;
        context = null;
        resources = null;
        invite = null;
        analytics = null;
        paymentHolder = null;
        broadcastBtcInviteHelper = null;
        transactionData = null;
        transactionHelper = null;
        externalNotificationHelper = null;
        walletHelper = null;
        broadcastTransactionHelper = null;
        inviteTransactionSummaryHelper = null;
        transactionNotificationManager = null;
        transactionBroadcastResult = null;
        transactionsInvitesSummary = null;
        syncWalletManager = null;
        phoneNumber = null;
        serverId = null;
    }

    @Before
    public void setUp() throws Exception {
        transactionsInvitesSummary = mock(TransactionsInvitesSummary.class);
        when(context.getResources()).thenReturn(resources);
        when(invite.getTransactionsInvitesSummary()).thenReturn(transactionsInvitesSummary);
        when(invite.getServerId()).thenReturn(serverId);
        when(invite.getReceiverPhoneNumber()).thenReturn(phoneNumber);
        String nameOfReceiver = "Joe Blow";
        when(invite.getInviteName()).thenReturn(nameOfReceiver);
        when(invite.getAddress()).thenReturn(address);
        long fee = 100L;
        when(invite.getValueFeesSatoshis()).thenReturn(fee);
        long amountToSend = 100000L;
        when(invite.getValueSatoshis()).thenReturn(amountToSend);
        USDCurrency evaluationCurrency = mock(USDCurrency.class);
        when(paymentHolder.getEvaluationCurrency()).thenReturn(evaluationCurrency);
        when(paymentHolder.getBtcCurrency()).thenReturn(new BTCCurrency(amountToSend));
        when(paymentHolder.getTransactionData()).thenReturn(transactionData);
        when(transactionData.getPaymentAddress()).thenReturn(address);
        transactionBroadcastResult = mock(TransactionBroadcastResult.class);
        when(broadcastTransactionHelper.broadcast(any())).thenReturn(transactionBroadcastResult);
        when(transactionBroadcastResult.isSuccess()).thenReturn(true);
        when(transactionFundingManger.buildFundedTransactionDataForDropBit(invite.getValueSatoshis(), invite.getValueFeesSatoshis())).thenReturn(transactionData);
        runner.setInvite(invite);
    }

    @Test
    public void full_funding_run_successful_test() {
        TransactionData transactionData = new TransactionData(new UnspentTransactionOutput[1],
                invite.getValueSatoshis(), invite.getValueFeesSatoshis(), 0, mock(DerivationPath.class), "");
        when(transactionFundingManger.buildFundedTransactionDataForDropBit(invite.getValueSatoshis(), invite.getValueFeesSatoshis())).thenReturn(transactionData);

        runner.run();

        assertThat(transactionData.getPaymentAddress(), equalTo(address));
        verify(broadcastTransactionHelper).broadcast(transactionData);
        verify(inviteTransactionSummaryHelper).updateFulfilledInvite(transactionsInvitesSummary, transactionBroadcastResult);
        verify(transactionNotificationManager).notifyCnOfFundedInvite(invite);
        verify(syncWalletManager).syncNow();
        verify(analytics).trackEvent(Analytics.EVENT_DROPBIT_COMPLETED);
    }

    @Test
    public void broadcast_TX_To_Btc_Network_error_test() {
        when(transactionData.getUtxos()).thenReturn(new UnspentTransactionOutput[1]);
        TransactionBroadcastResult result = mock(TransactionBroadcastResult.class);
        when(result.isSuccess()).thenReturn(false);
        when(broadcastTransactionHelper.broadcast(transactionData)).thenReturn(result);

        runner.run();

        verify(result).getMessage();
    }

    @Test
    public void on_error_funding_notify_server_cancel_transaction_test() {
        when(transactionData.getUtxos()).thenReturn(new UnspentTransactionOutput[0]);
        when(context.getString(anyInt(), anyString())).thenReturn("DropBit to 2565245258 has been canceled");

        runner.run();

        verify(externalNotificationHelper).saveNotification(eq("DropBit to 2565245258 has been canceled"), eq(serverId));
        verify(walletHelper).updateBalances();
        verify(walletHelper).updateSpendableBalances();
    }

    @Test
    public void on_error_funding_notify_user_that_the_transaction_is_canceled_test() {
        when(transactionData.getUtxos()).thenReturn(new UnspentTransactionOutput[0]);
        String cancelMessage = "Invite to " + invite.getReceiverPhoneNumber() + " has be canceled";

        when(context.getString(anyInt(), anyString())).thenReturn(cancelMessage);

        runner.run();
        verify(transactionHelper).updateInviteAsCanceled(eq(invite.getServerId()));

    }
}

File: app/src/test/java/com/coinninja/coinkeeper/service/runner/BroadcastBtcInviteRunnerTest.kt
package com.coinninja.coinkeeper.service.runner

import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.coinninja.bindings.DerivationPath
import com.coinninja.bindings.TransactionData
import com.coinninja.bindings.UnspentTransactionOutput
import com.coinninja.coinkeeper.bitcoin.BroadcastResult
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary
import com.coinninja.coinkeeper.model.db.TransactionsInvitesSummary
import com.coinninja.coinkeeper.util.analytics.Analytics
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verifyZeroInteractions
import com.nhaarman.mockitokotlin2.whenever
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.IsEqual.equalTo
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify

@RunWith(AndroidJUnit4::class)
class BroadcastBtcInviteRunnerTest {
    private fun createBroadcastInviteRunner(): BroadcastBtcInviteRunner {
        return BroadcastBtcInviteRunner(
                ApplicationProvider.getApplicationContext(),
                mock(), mock(), mock(), mock(), mock(), mock(), mock(), mock(), mock()
        )
    }

    @Test
    fun full_funding_run_successful_test() {
        val runner = createBroadcastInviteRunner()
        val transactionBroadcastResult = mock<BroadcastResult>()
        val txid = "-- txid --"
        whenever(transactionBroadcastResult.isSuccess).thenReturn(true)
        whenever(transactionBroadcastResult.txid).thenReturn(txid)
        runner.invite = mock()
        whenever(runner.invite!!.serverId).thenReturn("--server-id--")
        whenever(runner.invite!!.valueFeesSatoshis).thenReturn(100)
        whenever(runner.invite!!.valueSatoshis).thenReturn(10000)
        whenever(runner.invite!!.address).thenReturn("--address--")
        val transactionData = TransactionData(
                arrayOf(mock()),
                runner.invite!!.valueSatoshis,
                runner.invite!!.valueFeesSatoshis,
                0,
                mock(),
                ""
        )

        whenever(runner.transactionFundingManager.buildFundedTransactionDataForDropBit(runner.invite!!.address,
                runner.invite!!.valueSatoshis, runner.invite!!.valueFeesSatoshis)).thenReturn(transactionData)
        whenever(runner.broadcastHelper.broadcast(transactionData)).thenReturn(transactionBroadcastResult)

        runner.run()

        assertThat(transactionData.paymentAddress, equalTo<String>("--address--"))
        verify(runner.broadcastHelper).broadcast(transactionData)
        verify(runner.inviteTransactionSummaryHelper).updateFulfilledInvite(runner.invite!!, txid)
        verify(runner.transactionNotificationManager).notifyCnOfFundedInvite(runner.invite!!)
        verify(runner.syncWalletManager).syncNow()
        verify(runner.analytics).trackEvent(Analytics.EVENT_DROPBIT_COMPLETED)
    }

    @Test
    fun broadcast_TX_To_Btc_Network_error_test() {
        val runner = createBroadcastInviteRunner()
        runner.invite = mock()
        whenever(runner.invite!!.valueFeesSatoshis).thenReturn(100)
        whenever(runner.invite!!.valueSatoshis).thenReturn(10000)
        whenever(runner.invite!!.transactionsInvitesSummary).thenReturn(mock())
        whenever(runner.invite!!.address).thenReturn("--address--")
        val transactionData = TransactionData(arrayOf(mock()),
                runner.invite!!.valueSatoshis,
                runner.invite!!.valueFeesSatoshis,
                0, mock(), ""
        )
        whenever(runner.transactionFundingManager.buildFundedTransactionDataForDropBit(runner.invite!!.address,
                runner.invite!!.valueSatoshis, runner.invite!!.valueFeesSatoshis)).thenReturn(transactionData)

        val result = mock<BroadcastResult>()
        whenever(result.isSuccess).thenReturn(false)
        whenever(runner.broadcastHelper.broadcast(transactionData)).thenReturn(result)

        runner.run()

        verifyZeroInteractions(runner.inviteTransactionSummaryHelper)
    }

    @Test
    fun on_error_funding_cancels_transaction_test() {
        val runner = createBroadcastInviteRunner()
        runner.invite = mock(InviteTransactionSummary::class.java)
        whenever(runner.invite!!.valueFeesSatoshis).thenReturn(100)
        whenever(runner.invite!!.valueSatoshis).thenReturn(10000)
        whenever(runner.invite!!.transactionsInvitesSummary).thenReturn(mock(TransactionsInvitesSummary::class.java))
        whenever(runner.invite!!.address).thenReturn("--address--")
        whenever(runner.invite!!.serverId).thenReturn("--server-id--")
        val transactionData = TransactionData(
                emptyArray(),
                runner.invite!!.valueSatoshis,
                runner.invite!!.valueFeesSatoshis,
                0,
                mock(DerivationPath::class.java),
                ""
        )
        whenever(runner.transactionFundingManager.buildFundedTransactionDataForDropBit(runner.invite!!.address,
                runner.invite!!.valueSatoshis, runner.invite!!.valueFeesSatoshis)).thenReturn(transactionData)

        whenever(runner.invite!!.localeFriendlyDisplayIdentityForReceiver).thenReturn("2565245258")

        runner.run()

        verify(runner.externalNotificationHelper).saveNotification(
                "DropBit to 2565245258 has been canceled", runner.invite!!.serverId)
        verify(runner.cnWalletManager).updateBalances()
        verify(runner.inviteTransactionSummaryHelper).cancelInviteByCnId(runner.invite!!.serverId)
    }
}